### PR TITLE
test: Add .NET 7 Unit Test Build Targets

### PR DESCRIPTION
**Pull request by tippmar-nr**
Thank you for submitting a pull request.  Please review our [contributing guidelines](/CONTRIBUTING.md) and [code of conduct](https://opensource.newrelic.com/code-of-conduct/).

## Description

Adds .NET 7 build targets for existing unit test projects and refactors as necessary to get the .NET 7 tests to run successfully.

This refactoring resulted in some changes to `CoreContainer` that facilitate functionality equivalent to `WindsorContainer` when replacing component registrations with mocks for testing. 

# Author Checklist
- [ ] Unit tests, Integration tests, and Unbounded tests completed
- [ ] Performance testing completed with satisfactory results (if required)

# Reviewer Checklist
- [ ] Perform code review
- [ ] Pull request was adequately tested (new/existing tests, performance tests)

### CODE REVIEW

File: .github/workflows/run_unit_tests.yml [hunk 1]
```diff
@@ -44,7 +44,7 @@ jobs:
         run: dotnet restore
 
       - name: Build & Run Unit Tests
-        run: dotnet test --verbosity minimal --no-restore --settings tests\UnitTests.runsettings --results-directory ${{ env.test_results_path }}
+        run: dotnet test --verbosity normal --no-restore --settings tests\UnitTests.runsettings --results-directory ${{ env.test_results_path }}
```
Comment (chynesNR) [inline .github/workflows/run_unit_tests.yml:47]: Do we still want this at `normal`? It looks like a difference of 20K lines at normal vs 1K at minimal

File: tests/Agent/UnitTests/CompositeTests/Samplers/GcSamplerNetCoreTests.cs [hunk 2]
```diff
@@ -75,75 +75,75 @@ public void SamplerStartsWithoutException()
             sampler.Start();
         }
 
-        [Test]
-        public void SamplerDisposesEventListenerOnException()
-        {
-            var samplerWasStopped = false;
-            var listenerWasDisposed = false;
+        //[Test]
```
Comment (chynesNR) [inline tests/Agent/UnitTests/CompositeTests/Samplers/GcSamplerNetCoreTests.cs:78]: Do we need to come back to these?

File: .github/workflows/run_unit_tests.yml (hunk 1 above)
Comment (tippmar-nr) [reply .github/workflows/run_unit_tests.yml:47]: I can probably revert it - needed it to see where things were failing.

File: src/Agent/NewRelic/Agent/Core/Samplers/GCSamplerNetCore.cs [hunk 3]
```diff
@@ -129,7 +129,7 @@ public override void Dispose()
         {
             base.Dispose();
             _listener?.StopListening();
-            _listener?.Dispose();
+            //_listener?.Dispose();
```
[the remark this answers is not in the document]
Comment (tippmar-nr) [reply src/Agent/NewRelic/Agent/Core/Samplers/GCSamplerNetCore.cs]: I verified that `CoreContainer` was never disposing the AutoFac container in our old code, which meant this method was never getting called either. My fix for now is to call `_listener?.Dispose()` only on NETFRAMEWORK build targets.

File: tests/Agent/UnitTests/CompositeTests/CrossAgentTests/DistributedTracing/TraceContextCrossAgentTests.cs [hunk 4]
```diff
@@ -60,7 +60,13 @@ private static List<TestCaseData> GetTraceContextTestData()
         {
             var testCaseData = new List<TestCaseData>();
 
-            var dllPath = Path.GetDirectoryName(new Uri(Assembly.GetExecutingAssembly().CodeBase).LocalPath);
+            string location;
+#if NETFRAMEWORK
```
[the remark this answers is not in the document]
Comment (tippmar-nr) [reply tests/Agent/UnitTests/CompositeTests/CrossAgentTests/DistributedTracing/TraceContextCrossAgentTests.cs:64]: Yep, I can do that.

File: tests/Agent/UnitTests/CompositeTests/Samplers/GcSamplerNetCoreTests.cs (hunk 2 above)
Comment (tippmar-nr) [reply tests/Agent/UnitTests/CompositeTests/Samplers/GcSamplerNetCoreTests.cs:78]: I'll include them for .NET framework builds

Comment (tippmar-nr) [reply tests/Agent/UnitTests/CompositeTests/Samplers/GcSamplerNetCoreTests.cs:78]: Actually, these tests pass on both Framework and .NET so I'll put them back in.

**Review by chynesNR — APPROVED** [review]
Looks good!

**Review by nrcventura — APPROVED** [review]

**Pull request merged**